### PR TITLE
chore!: remove Guava Range()

### LINKED ISSUE #511: Get rid of Guava
Do you really need guava as runtime dependency?

### PR DESCRIPTION
**Pull request by thinkingserious**
# Fixes #

Fixes #511 

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java [hunk 1]
```diff
@@ -321,20 +285,12 @@ private void addQueryParams(final Request request) {
...
-        } else if (rangeStartTime != null) {
-            request.addQueryDateTimeRange("StartTime", rangeStartTime);
-        }
-
-        if (absoluteEndTime != null) {
-            request.addQueryParam("EndTime", absoluteEndTime.toString(Request.QUERY_STRING_DATE_TIME_FORMAT));
-        } else if (rangeEndTime != null) {
-            request.addQueryDateTimeRange("EndTime", rangeEndTime);
+        if (startTime != null || endTime != null) {
+            request.addQueryDateTimeRange("Range", startTime, endTime);
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:289]: `"Range"` is not correct. Where is this coming from?

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java [hunk 2]
```diff
@@ -321,20 +285,12 @@ private void addQueryParams(final Request request) {
...
         }
 
-        if (absoluteStartTime != null) {
-            request.addQueryParam("StartTime", absoluteStartTime.toString(Request.QUERY_STRING_DATE_TIME_FORMAT));
-        } else if (rangeStartTime != null) {
-            request.addQueryDateTimeRange("StartTime", rangeStartTime);
-        }
-
-        if (absoluteEndTime != null) {
-            request.addQueryParam("EndTime", absoluteEndTime.toString(Request.QUERY_STRING_DATE_TIME_FORMAT));
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: This removes functionality where I could before provide a single `EndTime` rather than needing to provide both upper and lower bound on the end time.

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java (hunk 1 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:289]: This was a mistake, should be "EndTime", will fix.

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java (hunk 2 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: I will add a function that replaces that functionality.

Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: You can still call `addQueryParam` as above; however, calling `addQueryDateTimeRange` should be transparent to the caller since if only `endTime` is defined, we achieve the same effect as calling `addQueryParam`. 

Are you proposing that we create an additional function that sits on top of `addQueryParam`, something along the lines of `addQueryEndDateTime`? Or in these cases, just use `addQueryParam` as before?

Comment (childish-sambino) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: No, I'm saying I used to be able to call `setStartTime` and it would add the query param `StartTime=`. Now it does `EndTime>=` I'm guessing?

Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: [This](https://github.com/twilio/twilio-java/blob/a5f7797dd4fab4a5f2f0f26d755e6c0431c13538/src/test/java/com/twilio/http/RequestTest.java#L121) is the current behavior when we just set the `startTime` and [this example](https://github.com/twilio/twilio-java/blob/a5f7797dd4fab4a5f2f0f26d755e6c0431c13538/src/test/java/com/twilio/http/RequestTest.java#L130) for setting just the `endTime`.

Comment (childish-sambino) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: Those are `Request` examples. I'm talking about the `CallReader` interface.

Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: Ah, how about replacing `setStartTime()` and `setEndTime()` with `setAbsoluteStartTime()`, `setAbsoluteEndTime()`, `setRangeStartTime()` and `setRangeEndTime()`?

Comment (childish-sambino) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: Why can't we use what all the other languages use? Examples: https://www.twilio.com/docs/sms/api/message-resource?code-sample=code-read-list-messages-filter-by-before-sent-date&code-language=Ruby&code-sdk-version=5.x

`setStartTime`, `setStartTimeBefore`, `setStartTimeAfter`.

Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: The biggest difference is that in other languages we can pass named parameters. In the closest implementation ([C#](https://github.com/twilio/twilio-csharp/blob/main/src/Twilio/Rest/Api/V2010/Account/CallResource.cs#L485)), we have [a builder](https://github.com/twilio/twilio-csharp/blob/main/src/Twilio/Rest/Api/V2010/Account/CallOptions.cs#L457) that sorts the date/time stuff.

In the Java case, we can keep the building of the query params in [`addQueryParams`](https://github.com/twilio/twilio-java/blob/main/src/main/java/com/twilio/rest/api/v2010/account/CallReader.java#L324) and then add the variables `startTime`, `startTimeBefore`, `startTimeAfter`, `endTimeBefore`, `endTime` and `endTimeAfter` along with their getters & setters. How does that implementation sound?

Comment (childish-sambino) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:331]: That's fine. As long as users can still individually provide `startTime`, `startTimeBefore`, and `startTimeAfter`.

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java [hunk 3]
```diff
@@ -127,46 +131,21 @@ public CallReader setStatus(final Call.Status status) {
...
         return this;
     }
 
-    /**
-     * Only include calls that started on this date. Specify a date as `YYYY-MM-DD`
-     * in GMT, for example: `2009-07-06`, to read only calls that started on this
-     * date. You can also specify an inequality, such as `StartTime&lt;=YYYY-MM-DD`,
-     * to read calls that started on or before midnight of this date, and
-     * `StartTime&gt;=YYYY-MM-DD` to read calls that started on or after midnight of
-     * this date..
```
Comment (eshanholtz) [inline src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:145]: What happened to these docs? They provide context for the difference between startTime and startTimeBefore, etc.

Comment (childish-sambino) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:145]: They're up above on `setStartTime`. It doesn't _really_ make sense on any of these methods specifically since it's talking about `StartTime&gt;=YYYY-MM-DD` and the like.

File: src/main/java/com/twilio/rest/api/v2010/account/MessageReader.java [hunk 4]
```diff
@@ -95,31 +98,23 @@ public MessageReader setFrom(final String from) {
...
      * messages sent on or after midnight on a date..
      *
-     * @param absoluteDateSent Filter by date sent
+     * @param dateSent Filter by date sent
      * @return this
      */
-    public MessageReader setDateSent(final DateTime absoluteDateSent) {
-        this.rangeDateSent = null;
-        this.absoluteDateSent = absoluteDateSent;
+    public MessageReader setDateSent(final LocalDate dateSent) {
```
Comment (eshanholtz) [inline src/main/java/com/twilio/rest/api/v2010/account/MessageReader.java:104]: Why is this a `LocalDate` type when it previously had date_time precision?

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java (hunk 3 above)
Comment (eshanholtz) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:145]: I think some clarification that the `Before()` maps to `StartTime&lt;=YYYY-MM-DD` and `After` maps to `StartTime&gt;=YYYY-MM-DD` would be helpful in these situations. Yes the docs say that you _can_ specify the inequality, but they're not on the methods that actually allow you to use that functionality.

Comment (childish-sambino) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:145]: The doc string is coming from the api definition.

Comment (eshanholtz) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:145]: Yeah, but we control the building of these additional functions. We can make the inequality setter builders render a clarifying function docstring ourselves that is independent of api-defs

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java [hunk 5]
```diff
@@ -127,46 +131,21 @@ public CallReader setStatus(final Call.Status status) {
...
-     *
-     * @param rangeStartTime Only include calls that started on this date
-     * @return this
-     */
-    public CallReader setStartTime(final Range<DateTime> rangeStartTime) {
-        this.absoluteStartTime = null;
-        this.rangeStartTime = rangeStartTime;
+    public CallReader setStartTimeBefore(final ZonedDateTime startTimeBefore) {
+        this.startTimeBefore = startTimeBefore;
```
[the remark this answers is not in the document]
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:143]: Ah, same for the other methods too, correct? I'll add some tests to verify this functionality.

File: src/main/java/com/twilio/rest/api/v2010/account/MessageReader.java (hunk 4 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/MessageReader.java:104]: Great catch! I'm going to double check all of these mappings. This particular one should be a `ZonedDateTime`.  

Reference:
`org.joda.time.DateTime ~> java.time.ZonedDateTime`
`org.joda.time.LocalDate ~> java.time.LocalDate`

Comment (eshanholtz) [reply src/main/java/com/twilio/rest/api/v2010/account/MessageReader.java:104]: Are you making use of the type mappings defined in api-defs? https://code.hq.twilio.com/twilio/api-definitions/blob/master/types/java.json

Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/MessageReader.java:104]: Yes: https://code.hq.twilio.com/twilio/api-definitions/pull/2068/files

This will also be updated as part of this PR to remove the Range type.

File: src/main/java/com/twilio/rest/api/v2010/account/CallReader.java (hunk 3 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/CallReader.java:145]: I will be looking at these comments again after implementing the Yoyodyne changes to ensure they make sense in context.

File: src/main/java/com/twilio/rest/api/v2010/account/Recording.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
 import com.twilio.rest.Domains;
 import lombok.ToString;
 
+
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/api/v2010/account/Recording.java:30]: Lots of trivial whitespace changes cluttering up the PR should be removed (to files where there are no other changes).

File: src/main/java/com/twilio/type/PhoneNumberPrice.java [hunk 7]
```diff
@@ -8,6 +8,8 @@
 
 import java.util.Objects;
 
+import lombok.ToString;
```
Comment (childish-sambino) [inline src/main/java/com/twilio/type/PhoneNumberPrice.java:11]: Duplicate imports in several files.

File: src/main/java/com/twilio/rest/events/v1/Schema.java [hunk 8]
```diff
@@ -109,7 +111,7 @@ private Schema(@JsonProperty("id")
         this.id = id;
         this.url = url;
         this.links = links;
-        this.lastCreated = DateConverter.iso8601DateTimeFromString(lastCreated);
+        this.lastCreated = ZonedDateTime.parse(lastCreated);
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/events/v1/Schema.java:114]: This should be in a separate PR.

File: src/main/java/com/twilio/rest/api/v2010/account/call/NotificationReader.java [hunk 9]
```diff
@@ -218,14 +215,12 @@ private void addQueryParams(final Request request) {
             request.addQueryParam("Log", log.toString());
         }
 
-        if (absoluteMessageDate != null) {
-            request.addQueryParam("MessageDate", absoluteMessageDate.format(DateTimeFormatter.ofPattern(Request.QUERY_STRING_DATE_FORMAT)));
-        } else if (rangeMessageDate != null) {
-            request.addQueryDateRange("MessageDate", rangeMessageDate);
+        if (startMessageDate != null || endMessageDate != null) {
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/api/v2010/account/call/NotificationReader.java:218]: Why is there no `messageDate` here? Was this not generated by yoyodyne?

File: src/main/java/com/twilio/rest/api/v2010/account/RecordingReader.java [hunk 10]
```diff
@@ -225,10 +220,12 @@ public RecordingReader setConferenceSid(final String conferenceSid) {
...
      */
     private void addQueryParams(final Request request) {
-        if (absoluteDateCreated != null) {
-            request.addQueryParam("DateCreated", absoluteDateCreated.format(DateTimeFormatter.ofPattern(Request.QUERY_STRING_DATE_TIME_FORMAT)));
-        } else if (rangeDateCreated != null) {
-            request.addQueryDateTimeRange("DateCreated", rangeDateCreated);
+        if (dateCreated != null) {
+            request.addQueryParam("DateCreated", dateCreated.format(DateTimeFormatter.ofPattern(Request.QUERY_STRING_DATE_TIME_FORMAT)));
+        } else {
+            if (dateCreatedBefore != null || dateCreatedAfter != null) {
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/api/v2010/account/RecordingReader.java:226]: nit: Combine into `else if`.

File: src/main/java/com/twilio/rest/api/v2010/account/Recording.java (hunk 6 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/Recording.java:30]: Fixing these with the Yoyodyne generation.

File: src/main/java/com/twilio/type/PhoneNumberPrice.java (hunk 7 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/type/PhoneNumberPrice.java:11]: Fixing these with the Yoyodyne generation.

File: src/main/java/com/twilio/rest/api/v2010/account/call/NotificationReader.java (hunk 9 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/api/v2010/account/call/NotificationReader.java:218]: This is a manual mistake, fixing.

File: src/main/java/com/twilio/http/Request.java [hunk 11]
```diff
@@ -216,16 +216,17 @@ public void addQueryDateRange(final String name, final Range<LocalDate> range) {
...
      *
      * @param name  name of query parameter
-     * @param range date range
+     * @param lowerBound date lowerBound
+     * @param upperBound date upperBound
      */
-    public void addQueryDateTimeRange(final String name, final Range<ZonedDateTime> range) {
-        if (range.hasLowerBound()) {
-            String value = range.lowerEndpoint().withZoneSameInstant(ZoneOffset.UTC).format(DateTimeFormatter.ofPattern(QUERY_STRING_DATE_TIME_FORMAT));
+    public void addQueryDateTimeRange(final String name, final ZonedDateTime lowerBound, final ZonedDateTime upperBound) {
```
Comment (childish-sambino) [inline src/main/java/com/twilio/http/Request.java:222]: This is being called as `request.addQueryDateTimeRange("StartTime", startTimeBefore, startTimeAfter);`, but `startTimeBefore` is really an _upper_ bound.

Given that this was able to pass without any tests breaking, tests need to be updated to catch this.

**Review by childish-sambino — CHANGES_REQUESTED** [review]

File: src/main/java/com/twilio/rest/events/v1/Schema.java (hunk 8 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/rest/events/v1/Schema.java]: Thought I caught that one ...

File: src/main/java/com/twilio/http/Request.java (hunk 11 above)
Comment (thinkingserious) [reply src/main/java/com/twilio/http/Request.java:222]: Nice catch! Will fix.

File: src/main/java/com/twilio/http/Request.java [hunk 12]
```diff
@@ -198,16 +197,27 @@ private String buildURL() {
...
-     * @param range date range
+     * @param lowerBound lower bound of LocalDate range
+     * @param upperBound upper bound of LocalDate range
      */
-    public void addQueryDateRange(final String name, final Range<LocalDate> range) {
-        if (range.hasLowerBound()) {
-            String value = range.lowerEndpoint().format(DateTimeFormatter.ofPattern(QUERY_STRING_DATE_FORMAT));
+    public void addQueryDateRange(final String name, LocalDate lowerBound, LocalDate upperBound) {
+        // automatically switch bounds if they were passed in backwards
+        if (lowerBound != null && upperBound != null) {
```
Comment (childish-sambino) [inline src/main/java/com/twilio/http/Request.java:205]: We don't do this in other languages. I'm against doing it here.

**Review by childish-sambino — APPROVED** [review]
